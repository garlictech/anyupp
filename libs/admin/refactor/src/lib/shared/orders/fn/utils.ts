import { CrudSdk } from '@bgap/crud-gql/api';

export interface OrderHandlerDeps {
  crudSdk: CrudSdk;
  timestamp: () => number;
}
