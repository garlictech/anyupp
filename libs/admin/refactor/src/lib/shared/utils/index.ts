import { ImageCompressorService } from './services/image-compressor/image-compressor.service';
import { QrGeneratorService } from './services/qr-generator/qr-generator.service';
import { ToasterService } from './services/toaster/toaster.service';

export * from './const';
export * from './enums';
export * from './fn';
export { ImageCompressorService, ToasterService, QrGeneratorService };
