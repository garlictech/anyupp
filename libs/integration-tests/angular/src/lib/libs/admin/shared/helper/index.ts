import { Auth } from 'aws-amplify';

import { awsConfig } from '@bgap/crud-gql/api';
import {
  getCognitoUsername,
  testAdminUserPassword,
} from '@bgap/shared/fixtures';

export const signInToCognito = async () => {
  Auth.configure({
    ...awsConfig,
    authenticationFlowType: 'USER_PASSWORD_AUTH',
    aws_appsync_authenticationType: 'AMAZON_COGNITO_USER_POOLS',
  });

  await Auth.signIn(getCognitoUsername('monad'), testAdminUserPassword);
};

export const signOutFromCognito = async () => {
  await Auth.signOut();
};
