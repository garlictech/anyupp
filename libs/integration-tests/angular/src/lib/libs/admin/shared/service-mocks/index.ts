export class MockNbDialogService {
  open() {
    /** */
  }
}

export class MockTranslateService {
  get() {
    /** */
  }
}
