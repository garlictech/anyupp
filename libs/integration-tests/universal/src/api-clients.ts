import {
  awsConfig,
  getCrudSdkForIAM,
  getCrudSdkForUserPool,
} from '@bgap/crud-gql/api';
import {
  getAnyuppSdkForIAM,
  getAnyuppSdkForUserPool,
} from '@bgap/anyupp-gql/api';
import { Auth } from 'aws-amplify';
import { from } from 'rxjs';
import { map } from 'rxjs/operators';

Auth.configure({
  ...awsConfig,
  // See: https://github.com/aws-amplify/amplify-js/issues/6552#issuecomment-682259256
  authenticationFlowType: 'USER_PASSWORD_AUTH',
  aws_appsync_authenticationType: 'AMAZON_COGNITO_USER_POOLS',
});

export const createAuthenticatedCrudSdk = (
  userName: string,
  password: string,
) => from(Auth.signIn(userName, password)).pipe(map(getCrudSdkForUserPool));

export const createAuthenticatedAnyuppSdk = (
  userName: string,
  password: string,
) => from(Auth.signIn(userName, password)).pipe(map(getAnyuppSdkForUserPool));

export const createIamCrudSdk = () =>
  getCrudSdkForIAM(
    process.env.AWS_ACCESS_KEY_ID || '',
    process.env.AWS_SECRET_ACCESS_KEY || '',
  );

export const createIamAnyuppSdk = () =>
  getAnyuppSdkForIAM(
    process.env.AWS_ACCESS_KEY_ID || '',
    process.env.AWS_SECRET_ACCESS_KEY || '',
  );
