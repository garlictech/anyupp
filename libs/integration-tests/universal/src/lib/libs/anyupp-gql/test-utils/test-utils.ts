export const getSortedProductCatIds = (
  input: Array<{ productCategoryId: string }>,
) => input.map(x => x.productCategoryId).sort((a, b) => (a > b ? 1 : -1));
