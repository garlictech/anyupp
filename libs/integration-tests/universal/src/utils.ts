import * as R from 'ramda';
import { pipe } from 'fp-ts/lib/function';

export const ES_DELAY = 3000; //ms
export const DYNAMO_DELAY = 1000; //ms

// Replace UUID-s with a static string for jest matchers
export const maskV4UuidIds = <T>(data: T) =>
  pipe(
    data,
    JSON.stringify,
    R.replace(
      /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/g,
      'MASKED ID',
    ),
    (d: string) => JSON.parse(d) as T,
  );
