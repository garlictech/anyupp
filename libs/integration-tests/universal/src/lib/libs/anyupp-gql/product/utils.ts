import * as CrudApi from '@bgap/crud-gql/api';
import { throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { throwIfEmptyValue } from '@bgap/shared/utils';

export const getChainProduct = (sdk: CrudApi.CrudSdk, productId: string) => {
  return sdk.GetChainProduct({ id: productId }).pipe(
    throwIfEmptyValue<CrudApi.ChainProduct>(),
    catchError(err => {
      return throwError('ChainProdcuct is missing');
    }),
  );
};

export const getGroupProduct = (sdk: CrudApi.CrudSdk, productId: string) => {
  return sdk.GetGroupProduct({ id: productId }).pipe(
    throwIfEmptyValue<CrudApi.GroupProduct>(),
    catchError(err => {
      return throwError('GroupProdcuct is missing');
    }),
  );
};

export const getUnitProduct = (sdk: CrudApi.CrudSdk, productId: string) => {
  return sdk.GetUnitProduct({ id: productId }).pipe(
    throwIfEmptyValue<CrudApi.UnitProduct>(),
    catchError(err => {
      return throwError('UnitProdcuct is missing');
    }),
  );
};
