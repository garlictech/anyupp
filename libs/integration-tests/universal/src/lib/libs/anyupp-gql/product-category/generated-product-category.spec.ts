import * as CrudApi from '@bgap/crud-gql/api';
import {
  generatedProductCategoryFixture,
  generatedProductFixture,
  productCategoryFixture,
  testIdPrefix,
} from '@bgap/shared/fixtures';
import { RequiredId } from '@bgap/shared/types';
import { sortById } from '@bgap/shared/utils';
import {
  deleteGeneratedProductCategoriesForAUnit,
  listGeneratedProductCategoriesForUnits,
  reGenerateActiveProductCategoriesForAUnit,
} from '@bgap/backend/product-categories';
import { of } from 'rxjs';
import { delay, map, switchMap, tap } from 'rxjs/operators';
import { createIamCrudSdk } from '../../../../api-clients';
import { getSortedProductCatIds } from '../test-utils/test-utils';

const DYNAMODB_OPERATION_DELAY = 3000;
const TEST_NAME = 'GEN_PRODUCT_CAT';

const unitId = `${testIdPrefix}${TEST_NAME}_UNIT_ID`;

const productCategory_01: RequiredId<CrudApi.CreateProductCategoryInput> = {
  ...productCategoryFixture.productCategoryBase,
  id: `${testIdPrefix}${TEST_NAME}_01`,
};
const productCategory_02: RequiredId<CrudApi.CreateProductCategoryInput> = {
  ...productCategoryFixture.productCategoryBase,
  id: `${testIdPrefix}${TEST_NAME}_02`,
};
const productCategory_03_wont_be_regenerated: RequiredId<CrudApi.CreateProductCategoryInput> =
  {
    ...productCategoryFixture.productCategoryBase,
    id: `${testIdPrefix}${TEST_NAME}_03_WONT_BE_REGENERATED`,
  };
const productCategory_04: RequiredId<CrudApi.CreateProductCategoryInput> = {
  ...productCategoryFixture.productCategoryBase,
  id: `${testIdPrefix}${TEST_NAME}_04`,
};
const generatedProductCategory_03_wont_be_regenerated =
  generatedProductCategoryFixture.getGeneratedProductCategory(
    // because productNum === 0
    {
      id: productCategory_03_wont_be_regenerated.id, // THE id and the productCategoryId is not equal but in this test case it is NOT important
      unitId,
      productCategoryId: productCategory_03_wont_be_regenerated.id,
    },
  );

// const PRODUCT_NUM_FOR_BATCH_CRUD = 26; // should be > 25 because the batchSize is 25
// const productIds = [...Array(PRODUCT_NUM_FOR_BATCH_CRUD).keys()]
//   .map(id => id.toString().padStart(2, '0'))
//   .map(id => `${testIdPrefix}${TEST_NAME}ID_${id}`);

const generatedProducts: Array<CrudApi.CreateGeneratedProductInput> = [
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_01.id,
  }),
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_02.id,
  }),
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_02.id,
  }),
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_02.id,
  }),
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_02.id,
  }),
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_02.id,
  }),
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_04.id,
  }),
  generatedProductFixture.getGeneratedProduct({
    id: 'NotImportant',
    unitId,
    productCategoryId: productCategory_04.id,
  }),
];

describe('GenerateProductCategory tests', () => {
  const deps = {
    crudSdk: createIamCrudSdk(),
  };

  // it('should NOT fail without any product as Input', done => {
  //   reGenerateActiveProductCategoriesForAUnit(deps)([]).subscribe({
  //     next() {
  //       // SHOULD NOT TIMEOUT so this next callback should be triggered
  //       done();
  //     },
  //   });
  // });

  describe('complex tests', () => {
    const cleanup = () =>
      of('cleanup').pipe(
        // CleanUP
        switchMap(() => deleteGeneratedProductCategoriesForAUnit(deps)(unitId)),
        switchMap(() =>
          deps.crudSdk.DeleteProductCategory({
            input: { id: productCategory_01.id },
          }),
        ),
        switchMap(() =>
          deps.crudSdk.DeleteProductCategory({
            input: { id: productCategory_02.id },
          }),
        ),
        switchMap(() =>
          deps.crudSdk.DeleteProductCategory({
            input: { id: productCategory_03_wont_be_regenerated.id },
          }),
        ),
        switchMap(() =>
          deps.crudSdk.DeleteProductCategory({
            input: { id: productCategory_04.id },
          }),
        ),
      );

    beforeEach(async () => {
      await cleanup()
        .pipe(
          delay(DYNAMODB_OPERATION_DELAY),
          // Seeding
          switchMap(() =>
            deps.crudSdk.CreateProductCategory({
              input: productCategory_01,
            }),
          ),
          switchMap(() =>
            deps.crudSdk.CreateProductCategory({
              input: productCategory_02,
            }),
          ),
          switchMap(() =>
            deps.crudSdk.CreateProductCategory({
              input: productCategory_03_wont_be_regenerated,
            }),
          ),
          switchMap(() =>
            deps.crudSdk.CreateProductCategory({
              input: productCategory_04,
            }),
          ),
          // CREATE one PRE existing generated Product Category that won't be regenerated, so it should be deleted during the process
          switchMap(() =>
            deps.crudSdk.CreateGeneratedProductCategory({
              input: generatedProductCategory_03_wont_be_regenerated,
            }),
          ),
          delay(DYNAMODB_OPERATION_DELAY),
        )
        .toPromise();
    }, 25000);

    afterAll(() => cleanup().toPromise(), 15000);

    const startStateCheck = () =>
      of('GENERATE_PRODUCT_CATEGORIES').pipe(
        switchMap(() => listGeneratedProductCategoriesForUnits(deps)([unitId])),
        tap({
          next(result) {
            // ONLY THE PRECREATED GEN_CATEGORY SHOULD EXIST
            expect(getSortedProductCatIds(result)).toEqual([
              productCategory_03_wont_be_regenerated.id,
            ]);
          },
        }),
        delay(DYNAMODB_OPERATION_DELAY),
      );

    it('should generate the given product Categories with correct productNums or clear all in case the input is empty', done => {
      startStateCheck()
        .pipe(
          // PHASE 1: the pre existing 3. should be deleted and the new 1, 2, 4, should be generated with proper productNums
          switchMap(() =>
            reGenerateActiveProductCategoriesForAUnit(deps)({
              unitId,
              generatedProducts,
            }),
          ),
          delay(DYNAMODB_OPERATION_DELAY),
          switchMap(() =>
            listGeneratedProductCategoriesForUnits(deps)([unitId]),
          ),
          map(sortById),
          tap({
            next(result) {
              // ONLY THE PRECREATED GENCATEGORY SHOULD EXIST
              expect(getSortedProductCatIds(result)).toEqual([
                productCategory_01.id,
                productCategory_02.id,
                productCategory_04.id,
              ]);
              expect(result[0].productCategoryId).toEqual(
                productCategory_01.id,
              );
              expect(result[0].productNum).toEqual(1);
              expect(result[1].productCategoryId).toEqual(
                productCategory_02.id,
              );
              expect(result[1].productNum).toEqual(5);
              expect(result[2].productCategoryId).toEqual(
                productCategory_04.id,
              );
              expect(result[2].productNum).toEqual(2);
              expect(result[0]).toMatchSnapshot(
                {
                  createdAt: expect.any(String),
                  updatedAt: expect.any(String),
                  productCategory: {
                    createdAt: expect.any(String),
                    updatedAt: expect.any(String),
                  },
                },
                'A Generated ProductCategory',
              );
            },
          }),
          // PHASE 2: in case of an empty array as input all the existing generatedProductCategories should be deleted for the unit
          switchMap(() =>
            reGenerateActiveProductCategoriesForAUnit(deps)({
              unitId,
              generatedProducts: [],
            }),
          ),
          delay(DYNAMODB_OPERATION_DELAY),
          switchMap(() =>
            listGeneratedProductCategoriesForUnits(deps)([unitId]),
          ),
          tap({
            next(result) {
              expect(getSortedProductCatIds(result)).toEqual([]);
            },
          }),
        )
        .subscribe({
          next() {
            // SHOULD NOT TIMEOUT so this next callback should be triggered
            done();
          },
          error(err) {
            console.error(`${TEST_NAME}Test ERROR`, err);
          },
        });
    }, 60000);
  });
});
