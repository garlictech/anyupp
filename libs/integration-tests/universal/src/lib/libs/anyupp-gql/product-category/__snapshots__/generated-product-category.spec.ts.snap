// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateProductCategory tests complex tests should generate the given product Categories with correct productNums or clear all in case the input is empty: A Generated ProductCategory 1`] = `
Object {
  "createdAt": Any<String>,
  "id": "test_GEN_PRODUCT_CAT_UNIT_ID__test_GEN_PRODUCT_CAT_01",
  "position": 2,
  "productCategory": Object {
    "chainId": "seeded_chain_1_id",
    "createdAt": Any<String>,
    "description": Object {
      "de": null,
      "en": "DESCRIPTION",
      "hu": null,
    },
    "id": "test_GEN_PRODUCT_CAT_01",
    "image": "IMAGE",
    "name": Object {
      "de": null,
      "en": "PRODUCT_CATEGORY",
      "hu": null,
    },
    "parentId": null,
    "position": 0,
    "updatedAt": Any<String>,
  },
  "productCategoryId": "test_GEN_PRODUCT_CAT_01",
  "productNum": 1,
  "unitId": "test_GEN_PRODUCT_CAT_UNIT_ID",
  "updatedAt": Any<String>,
}
`;
